Generic reconciler

joamaki opened the pull request:
Add the generic reconciler along with example and benchmark applications.

joamaki committed:
hive/job: Add support for infinite retries for OneShot jobs

Allow specifying a negative retry count for infinite retries of
OneShot jobs.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
hive/job: Speed up tests with short rate-limiting and parallelism

The tests took 16s before, now down to less than a second.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
hive/cell: Add SetReporterMinTimeoutForTest for setting the minimum realization time

The 500ms default is way too long for tests that validate health status multiple times.
Add a function to allow overriding this.

TBD whether we want to keep this or whether to use a fake instead or some other
approach.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
reconciler: Add the statedb reconciler utility

This implements a generic utility for reconciling between
a statedb table and a target defined as a set of idempotent
operations. It implements incremental reconciliation with
per-object retries and a periodic full reconciliation
(forced per-object update and pruning).

A single shared reconciler implementation helps building a
more resilient architecture as it centralizes the decisions
on how to deal with failures and how to report them.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
reconciler: Add an example application

Example application for showing how to use the reconciler
to reconcile set of files on disk ("memos") created through
an HTTP API.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd commented on pkg/hive/cell/structured.go:
```diff
@@ -37,6 +37,12 @@ type children sets.Set[string]
 var reporterMinTimeout = time.Millisecond * 500
 
+// SetReporterMinTimeoutForTest sets the reporter minimum timeout to speed up validating
+// the health reporting when testing a module. Only to be used from tests.
```
could take a `testing.T` to enfore use only from tests, if we want to keep this

joamaki replied:
good idea, added and included cleanup too.

bimmlerd commented on pkg/hive/job/job.go:
```diff
@@ -265,6 +265,15 @@ func WithRetry(times int, backoff workqueue.RateLimiter) jobOneShotOpt {
 
+// WithAlwaysRetry retries a one shot job forever, even when no error is
+// returned.
+func WithAlwaysRetry() jobOneShotOpt {
```
Is this equivalent to a timer without a timeout? or does this get rate-limited? if so, maybe add a comment about that.

joamaki replied:
Ended up removing this and just adding comment to `WithRetry` that with negative retry count it'll retry forever. This way the rate-limiter can be set. Also realized the retries for loop() were not needed anymore after I dropped the Init() operation. I'll still keep the patch as it's useful.

bimmlerd commented on pkg/hive/job/job_test.go:
```diff
@@ -196,7 +196,7 @@ func testOneShot_RetryBackoff() (bool, error) {
 			// Test that the rate of change is 2 +- 50%, the 50% to account for CI time dilation.
```
comment not in sync with change

joamaki replied:
fixed!

bimmlerd commented on pkg/statedb/types.go:
```diff
@@ -57,12 +57,21 @@ type Table[Obj any] interface {
 	LowerBound(ReadTxn, Query[Obj]) (iter Iterator[Obj], watch <-chan struct{})
 
+	// NumObjects returns the number of objects stored in the table.
+	NumObjects(ReadTxn) int
```
nit: some alternate name possibilities: `Len` for Golang consitency, or `Rows()` for the table metaphor. unconvinced by all of them, though.

joamaki replied:
Yeah. Initially had `Length`, then changed it to `Size` and then finally to `NumObjects`. Perhaps could just go with `Len` just like iradix...

bimmlerd commented on pkg/statedb/types.go:
```diff
@@ -57,12 +57,21 @@ type Table[Obj any] interface {
 	DeleteTracker(txn WriteTxn, trackerName string) (*DeleteTracker[Obj], error)
+
+	// PrimaryKey returns the raw primary key for the object.
+	PrimaryKey(Obj) []byte
```
Somewhat breaks the API for me; reads like two ways of doing the same thing. Am I supposed to use `Get` or use this. If necessary, I'd rather see the table return a ref to its primary index, than exposing these.
The reason for these is mentioned in the commit message, but it's not immediately clear to me why the reconciler needs these, nor why this is better than providing the reconciler with the index. Could you elaborate?

joamaki replied:
I also feel the same way (which is why I added this as TODO). This was a bit of laziness from me and the desire to keep the dependencies to the reconciler simpler. I'll take a look at instead passing the primary indexer to it (either directly or having `PrimaryIndexer() Indexer[Obj]` in `Table[T]` and refactoring the `retries.go` to deal with that.

joamaki replied:
Ok, did a pass to build it around the Indexer now. PTAL.

bimmlerd replied:
Works for me now. only nit is that we now duplicate the primary indexer in the genTable, but I can live with that.

bimmlerd commented on pkg/hive/job/job_test.go:
```diff
@@ -102,6 +102,7 @@ func TestOneShot_RetryFail(t *testing.T) {
 	)
 
 	const retries = 3
+	rateLimiter := workqueue.NewItemExponentialFailureRateLimiter(10*time.Millisecond, 20*time.Millisecond)
```
if we're speeding up tests, can these run in parallel?

bimmlerd commented on pkg/statedb/reconciler/cell.go:
```diff
@@ -0,0 +1,13 @@
+
+var Cell = cell.Module(
+	"reconciler",
+	"Generic reconciler",
```
nit: could use some love

joamaki replied:
Added comments.

bimmlerd commented on pkg/statedb/reconciler/cell.go:
```diff
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: Apache-2.0
```
commit message which explains why this is useful would be great

joamaki replied:
Improved the commit message. Still not super inspired though :)

bimmlerd replied:
good enough for me^^

bimmlerd commented on pkg/statedb/reconciler/index.go:
```diff
@@ -0,0 +1,52 @@
+
+// WaitForReconciliation blocks until all objects have been reconciled or the context
+// has cancelled.
+func WaitForReconciliation[Obj any, Table statedb.Table[Obj]](
```
what is this useful for? by the time this function returns, the result may already be out of date right? Unless you hold a write transaction to prevent changes, but then it'd block indefinitely if there are pending things, as you are now blocking yourself.

joamaki replied:
It's useful for synchronizing when needed. E.g. if you add some pending objects you can call `WaitForReconciliation` and when it returns you know that at least those objects have now been reconciled. Arguably we could also not include this and let the user write the queries to do this themselves as needed (perhaps with finer granularity). It was pretty useful in the benchmark I just added though... :)

bimmlerd replied:
clearly it is useful :grin: 
I guess I'm a bit worried about this blocking indefinitely if we have a high enough rate of churn that we never 'finish' reconciling. E.g. with a constant rate of say 5/sec and operations taking ~a second each, this would be stuck until the context expires. And it's unclear how to handle the context expiration, by default I guess people will just return the err (even though the thing they're interested in probably actually reconciled _and_ there was no error)...

bimmlerd commented on pkg/statedb/reconciler/types.go:
```diff
@@ -0,0 +1,135 @@
+)
+
+type Config[Obj any] struct {
+	FullReconcilationInterval time.Duration
```
nit: doc comments on these

bimmlerd commented on pkg/statedb/reconciler/types.go:
```diff
@@ -0,0 +1,147 @@
+	// reconciliation is done periodically by calling 'Update' on all objects.
+	//
+	// 'changed' is returned as true if an update was performed. This allows
+	// tracking whether full reconciliation catched an out-of-sync target state.
```
```suggestion
	// tracking whether full reconciliation caught an out-of-sync target state.
```

bimmlerd commented on pkg/statedb/reconciler/types.go:
```diff
@@ -0,0 +1,147 @@
+	// reconciliation is performed when the desired state is updated. A full
+	// reconciliation is done periodically by calling 'Update' on all objects.
+	//
+	// 'changed' is returned as true if an update was performed. This allows
```
```suggestion
	// 'changed' must be true if an update was performed. This allows
```

joamaki replied:
Reworded.

bimmlerd commented on pkg/statedb/reconciler/types.go:
```diff
@@ -0,0 +1,147 @@
+	// tracking whether full reconciliation catched an out-of-sync target state.
+	Update(context.Context, statedb.ReadTxn, Obj) (changed bool, err error)
+
+	// Delete the object in the target. Same semantics as with Update.
```
```suggestion
	// Delete the object in the target. Same semantics as with Update.
	// Specifically, deleting a non-existent object is not an error.
```

joamaki replied:
Added.

bimmlerd commented on pkg/statedb/reconciler/types.go:
```diff
@@ -0,0 +1,147 @@
+	// Prune is called during full reconciliation after all desired objects
+	// have been validated with a call to Update().
+	// Prune finds objects in the target that are not in the set of desired
+	// objects and removes them. A no-op for some targets.
```
nit: 'targets' used here still

bimmlerd commented on pkg/statedb/reconciler/types.go:
```diff
@@ -0,0 +1,135 @@
+}
+
+type Reconciler[Obj any] interface {
+	TriggerFullReconciliation()
```
a note on how this behaves if called concurrently would be good

bimmlerd commented on pkg/statedb/reconciler/types.go:
```diff
@@ -0,0 +1,135 @@
+	Kind StatusKind
+
+	// Delete is true if the object should be deleted by the reconciler.
+	// If an object is deleted outside the reconciler it will not be
```
I don't understand what it means for an object to be deleted externally, could you elaborate?

joamaki replied:
e.g. someone uses `RWTable[Foo].Delete` directly instead of doing `Insert(Foo{..., Status: reconciler.StatusPendingDelete()})`. If this happens then the reconciler will not be aware of it and won't perform any operations on it. 

The reason why the reconciler uses soft deletes is really to be able to tell when a delete of an object failed and keep the object around until it's deletion has been reconciled successfully. This way we don't need a separate mechanism to tell that a delete is failing or a separate mechanism to wait for a reconciliation of a deletion.

Do you have a suggestion on how I'd document that here? There's some docs now in `StatusPendingDelete()` that I could refer the reader to here.

bimmlerd replied:
Ah- makes sense now that I've read the whole patch. I guess it's just something mildly subtle that's easy to miss, but hard to document better. Maybe it's enough to just give an example of external deletion (like you did in this thread).

bimmlerd commented on pkg/statedb/reconciler/reconciler.go:
```diff
@@ -0,0 +1,389 @@
+	process := func(key []byte, obj Obj, rev statedb.Revision) error {
+		start := time.Now()
+
+		status := r.Config.GetObjectStatus(obj)
```
why are we fetching the status again here? it seems to me that we just did, and hence got here - could pass it?

joamaki replied:
:+1:

bimmlerd commented on pkg/statedb/reconciler/reconciler.go:
```diff
@@ -0,0 +1,389 @@
+
+		if oldRev == newRev {
+			// No changes happened between the ReadTxn and this WriteTxn. Since
+			// we wrote the table the 'watch' channel has closed. Grab a new
```
I'm confused. This reads as if we only write if `oldRev == newRev`. But above it seems to me that we check the result.rev, which may be different, and hence a write could have succeeded even though `oldRev != newRev`, no?

joamaki replied:
`oldRev` is the revision that we just reconciled. It could be that something new happened after that and before we acquired the write transaction. If something did, then we know we want to reconcile again right away. If something didn't change then we can just watch changes after our write transaction. I reworked the code a little bit, perhaps it's clearer now?

bimmlerd replied:
yep that helped, nice!

bimmlerd commented on pkg/statedb/reconciler/reconciler.go:
```diff
@@ -0,0 +1,389 @@
+		float64(time.Since(start)) / float64(time.Second),
+	)
+
+	// Call Update() for each desired object to validate that it is up-to-date.
```
We did an incremental reconciliation (i.e. calling `Update` on all desired objects), then we pruned. Why are we calling update on all of them again?

joamaki replied:
That's what full reconciliation is about (full update of all objects + pruning of unexpected things). I guess we could maybe skip freshly incrementally reconciled, but I'm not sure it's worth the complexity.

bimmlerd replied:
ah, derp. yeah, somehow thought that incremental touches all, when of course it doesn't have to

bimmlerd commented on pkg/statedb/reconciler/reconciler_test.go:
```diff
@@ -0,0 +1,532 @@
+	ID_3 = uint64(3)
+)
+
+func TestReconciler(t *testing.T) {
```
Not currently `t.Parallel` safe - could make it so?

joamaki replied:
Yes if we can solve the `SetReporterMinTimeoutForTest` thing.

bimmlerd commented on pkg/statedb/reconciler/reconciler_test.go:
```diff
@@ -0,0 +1,532 @@
+func TestReconciler(t *testing.T) {
+	cell.SetReporterMinTimeoutForTest(10 * time.Millisecond)
+
+	defer goleak.VerifyNone(t)
```
I think in tests it's always preferrable to use `t.Cleanup` instead of defer.

joamaki replied:
That's not what `go doc goleak.VerifyNone` says.

bimmlerd replied:
Fair, but consider this: One of the subtle things about `t.Parallel()` is that it makes `t.Run` non-blocking. If you have a `defer` in the top-level test and mark _it_ as parallel, the deferred call will be executed concurrently to the parallel subtests. A cleanup call, on the other hand, will be executed after all subtests complete. 

Arguably, this isn't an issue if you only mark the subtests as parallel, but afaik there's no downside to `t.Cleanup` so it seems better to just use it.

Re `VerifyNone`'s doc comment, I'm not sure their point still stands if it's executed in a `Cleanup` - by then, any go routine leaked is a true leak I think? But it's unclear whether the issue with associating the goroutine to a specific test breaks it nonetheless.

Both `go doc`s reproduced below for ease of discussion.

```
func VerifyNone(t TestingT, options ...Option)
    VerifyNone marks the given TestingT as failed if any extra goroutines are
    found by Find. This is a helper method to make it easier to integrate in
    tests by doing:

        defer VerifyNone(t)

    VerifyNone is currently incompatible with t.Parallel because it cannot
    associate specific goroutines with specific tests. Thus, non-leaking
    goroutines from other tests running in parallel could fail this check.
    If you need to run tests in parallel, use VerifyTestMain instead, which will
    verify that no leaking goroutines exist after ALL tests finish.
```
```
func (c *T) Cleanup(f func())
    Cleanup registers a function to be called when the test (or subtest) and
    all its subtests complete. Cleanup functions will be called in last added,
    first called order.
```

bimmlerd commented on pkg/statedb/reconciler/reconciler_test.go:
```diff
@@ -0,0 +1,532 @@
+
+	numIterations := 10
+
+	t.Run("incremental", func(t *testing.T) {
```
this is already a pretty awesome test framework. I've thought about how easy it'd be to add a new test, and I think it's pretty good already. One thing that you could try, if so inclined, would be to use https://pkg.go.dev/rsc.io/script to make the DSL more explicit. That way, the test definition would _just_ be the script, without any of the setup, which I think might be a way to make it even easier to add new tests. (inspired by https://research.swtch.com/testing)

joamaki replied:
Hmm maybe. What I wanted was 1) make it easy to extend the DSL, 2) make sure stack traces made sense. My earlier variant had really bad stack traces as the test was composed as `[]testCase`. Not sure how it'd look like with `rsc.io/script`. Perhaps we can revisit if this gets more complicated?

bimmlerd replied:
Yeah, mostly an idea at this point. I've not yet use the script stuff in anger, and have no idea how the stack traces look. As is I'd say is already good enough for writing tests for the inevitable issues, if we need more we can invest then

bimmlerd commented on pkg/statedb/reconciler/reconciler_test.go:
```diff
@@ -0,0 +1,532 @@
+func assertSensibleMetricDuration(t *testing.T, metrics map[string]float64, metric string) {
+	assert.Less(t, metrics[metric], 1.0, "expected metric %q to be above zero", metric)
+
+	// TODO: Sometimes the histogram metric is 0.0 even though samples have been added. Figure
```
maybe there's something stochastic about the sampling?

joamaki replied:
yeah that's my guess as well. annoying though! might need to think about abstracting the metrics ops somehow...

bimmlerd commented on pkg/statedb/reconciler/retries.go:
```diff
@@ -0,0 +1,131 @@
+	numRetries int       // number of retries attempted (for calculating backoff)
+}
+
+// Wait returns a channel that is closed when there is an item to retry.
```
should add that if there is none at the time we call, blocking on this will block forever, even if retryable things are added to the queue

bimmlerd commented on pkg/statedb/reconciler/retries.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+// retries holds the items that failed to be reconciled in
+// a priority queue ordered by retry time.
```
maybe a note about this not being safe to concurrently access

joamaki replied:
since this is internal I'm not sure that's needed. will add it anyway.

bimmlerd reviewed (commented):
overall looks really good!

joamaki committed:
reconciler: Add benchmark program

Example run:

  benchmark % go build .
  benchmark % ./benchmark -objects=100000 -batchsize=1000
  Inserting batch 100/100 ...
  Waiting for reconciliation to finish ...

  100000 objects reconciled in 1.02 seconds (batch size 1000)
  Throughput 98225.12 objects per second

Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd commented on pkg/statedb/table.go:
```diff
@@ -116,6 +122,24 @@ func (t *genTable[Obj]) Revision(txn ReadTxn) Revision {
+	return t.primaryAnyIndexer.fromObject(object{data: obj}).First()
+}
+
+func (t *genTable[Obj]) GetByPrimaryKey(txn ReadTxn, key []byte) (obj Obj, revision uint64, ok bool) {
```
is this still needed now?

joamaki replied:
oops forgot to delete. fixed.

bimmlerd commented on pkg/statedb/reconciler/reconciler_test.go:
```diff
@@ -0,0 +1,538 @@
+		new := h.ops.updates.get(id)
+		return new > old
+	}
+	require.Eventually(h.t, cond, time.Second, time.Millisecond, "expected %d to be retried", id)
```
nit: this should be assert, I think (applies to some more instances).
I tend to use `assert` for test, and `require` for setup/things not under test
but not sure there's even a difference in the output, so :shrug:

joamaki replied:
I don't think it should be an assert.  We want to stop/fatal the test immediately as the test steps build upon the previous ones so it'll get really confusing if we keep going, e.g. only the first assertion failure makes any sense.

"Package require implements the same assertions as the `assert` package but stops test execution when a test fails. "

bimmlerd commented on pkg/statedb/reconciler/reconciler_test.go:
```diff
@@ -0,0 +1,538 @@
+	require.Eventually(h.t, cond, time.Second, time.Millisecond, "expected object %d to not be found", id)
+}
+
+func (h testHelper) expectOp(op string) {
```
either here or in the comment for the testhelper - how do I specify the ops, what's the language used. I can kind of infer it from the existing test cases, but would be great to have written down

joamaki replied:
it's pretty ad-hoc so one would need to just read the helper code. I would rather not document it as it's defined within few lines of code and IMO here it's better to read the code than read possibly out-of-date comments. I suppose I could pull out the strings into constructors, but I'm not sure if that buys much.

EDIT: hmm or maybe it does, I'll try it out.
EDIT2: yeah, I think it is better with constructors.

bimmlerd commented on pkg/statedb/reconciler/reconciler_test.go:
```diff
@@ -0,0 +1,538 @@
+			h.expectStatus(ID_2, reconciler.StatusKindError, "update fail")
+			h.expectStatus(ID_3, reconciler.StatusKindError, "update fail")
+
+			// Make the ops health again and check that the objects recover.
```
```suggestion
			// Make the ops healthy again and check that the objects recover.
```

bimmlerd commented on pkg/statedb/reconciler/reconciler_test.go:
```diff
@@ -0,0 +1,538 @@
+		}
+	})
+
+	t.Run("full", func(t *testing.T) {
```
This afaict doesn't cover the external deletion case, can we test that we detect out of sync things in that case?

joamaki replied:
All we care here is to test that we call it with the right arguments and that's what we're doing by asserting that prune is called with N objects. We can't really "cover the external deletion case" as there is no "external" here (as in no "realized state"). We're just checking what operations are called without at all modelling what is being reconciled.

the out-of-sync metric isn't tested very thoroughly though (just that it's non-zero at the end of the test).

joamaki replied:
... and the out-of-sync metric is broken if full reconciliation of objects fail and are retried. hmm not sure how to cleanly fix this. might need to keep track of whether the item was queued from incremental or full reconciliation. though one option is to just ignore the retry-case. the out-of-sync is mostly just a hint that something outside cilium-agent is messing with the realized state...

bimmlerd replied:
Yeah I guess it depends on the intended semantics of the metric - if we want to say 'how many times we encountered items which are out of sync' then I think it's fine if we count the same multiple times.

joamaki replied:
The expected use-case for the metric is to essentially look whether "outOfSync > 0" within some time span and then investigate further if that's the case. E.g. we'd use this to catch if systemd is messing with us.

bimmlerd commented on pkg/statedb/reconciler/reconciler_test.go:
```diff
@@ -0,0 +1,538 @@
+
+			// Full reconciliation with functioning ops.
+			t.Log("Full reconciliation with non-faulty ops")
+			h.triggerFullReconciliation()
```
maybe worth a comment _why_ we expect a full prune here
The test implementation of prune differs a bit from the expected implementation - the test indiscriminatorily deletes all, whereas a real prune should only delete unexpected, right?

joamaki replied:
I'm not sure I follow. what's a full prune? we're just asserting that prune gets called with N objects which define the desired state. the semantics of Prune are implementation-specific, and it's up to the implementation to figure out if there is something to be done based on the realized state.

joamaki replied:
I tried improving the documentation for Prune(), hopefully clearer now.

bimmlerd replied:
ah. well, I was misunderstanding the test DSL - I thought `prune: 3` means three things got pruned, not we called prune with three things. The latter of course makes more sense.

bimmlerd reviewed (approved):
some more nits, but this is in pretty good shape already :)

aanm commented on pkg/statedb/reconciler/types.go:
```diff
@@ -0,0 +1,207 @@
+	// retrying.
+	RetryBackoffMaxDuration time.Duration
+
+	// IncrementalBatchSize is the maximum number objects to reconcile during
+	// incremental reconciliation before updating status and refreshing the
+	// statedb snapshot. This should be tuned based on the cost of each operation
+	// and the rate of expected changes so that health and per-object status
+	// updates are not delayed too much. If in doubt, use a value between 100-1000.
+	IncrementalBatchSize int
```
This one is a little bit confusing to understand. As a user of this interface, it's not clear to me which value I should be using here. It almost looks like this is an internal implementation detail? :thinking:

joamaki replied:
Yeah, this was a bit of an afterthought. I think we'd probably be OK with hardcoding it. We'd only run into trouble if indvidual operations are actually very costly (like doing network roundtrips) and there if we get 1000s of changed objects then it might take a long time before the health and object status gets updated.

Also could use a better name, perhaps `IncrementalMaxObjectsPerRound` or something more like that. "Batch" is confusing.

aanm commented on pkg/statedb/reconciler/types.go:
```diff
@@ -0,0 +1,207 @@
+// and a ReadTxn to allow for the option of looking up realized state from another
+// statedb table as an optimization (main use-case is reconciling routes against
+// Table[Route] to avoid a syscall per route).
+type Operations[Obj any] interface {
```
For all of the operations, if the given `context.Context` is canceled, what should be the behavior for the implementations? Should the revert or should they simply "stop" that operation?

joamaki replied:
I've honestly not thought of its semantics much since all the reconcilers thus far have done just quick uninterruptible syscalls. The context is really just telling that hey we're shutting down and please try to stop. So it's probably mostly applicable for long-running cancellable operations like doing a REST API call or something. I think "simply stop" is likely most we can ask for here.

aanm commented on pkg/statedb/reconciler/types.go:
```diff
@@ -0,0 +1,207 @@
+	// Deleting a non-existing object is not an error and returns nil.
+	Delete(context.Context, statedb.ReadTxn, Obj) error
+
+	// Prune undesired state. It is given an iterator for the full set of
+	// desired objects. The implementation should diff the desired state against
+	// the realized state to find things to prune.
+	// Invoked during full reconciliation before the individual objects are Update()'d.
+	//
+	// Unlike failed Update()'s a failed Prune() operation is not retried until
+	// the next full reconciliation round.
+	Prune(context.Context, statedb.ReadTxn, statedb.Iterator[Obj]) error
```
What's the purpose of `statedb.ReadTxn` if we have `statedb.Iterator[Obj]`? For example the `Prune` implementation done in pkg/statedb/reconciler/example/ops.go, isn't using the `statedb.ReadTxn`.

joamaki replied:
The only use-case thus far has been in the route reconciler (not part of this PR though). There the operations could check against `Table[*Route]` (a mirror of the realized state) instead of doing syscalls. I'll make sure this gets documented!

joamaki replied:
Ah looks like I did document it as part of the doc comment of `Operations`
